pkg/audit/manager.go

```diff
@@ -0,0 +1,270 @@
+package audit
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"strings"
+	"time"
+
+	opa "github.com/open-policy-agent/frameworks/constraint/pkg/client"
+	"github.com/pkg/errors"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("controller").WithValues("metaKind", "audit")
+
+const (
+	crdName       = "constrainttemplates.templates.gatekeeper.sh"
+	constraintsGV = "constraints.gatekeeper.sh/v1alpha1"
+	msgSize       = 256
+)
+
+var (
+	auditInterval = flag.Int("auditInterval", 60, "interval to run audit in seconds. defaulted to 60 secs if unspecified ")
+)
+
+// auditManager allows us to audit resources periodically
+type auditManager struct {
+	mgr     manager.Manager
+	opa     opa.Client
+	stopper chan struct{}
+	stopped chan struct{}
+	cfg     *rest.Config
+}
+
+type auditResult struct {
+	cname       string
+	cgroup      string
+	ckind       string
+	cversion    string
+	capiversion string
+	rkind       string
+	rname       string
+	rnamespace  string
+	message     string
+}
+
+// New starts a new manager for audit
+func New(ctx context.Context, cfg *rest.Config, opa opa.Client) error {
+	am := &auditManager{
+		opa:     opa,
+		stopper: make(chan struct{}),
+		stopped: make(chan struct{}),
+		cfg:     cfg,
+	}
+	mgr, err := manager.New(cfg, manager.Options{})
+	if err != nil {
+		return err
+	}
+	am.mgr = mgr
+	go startMgr(mgr, am.stopper, am.stopped)
+	go am.auditManagerLoop(ctx)
+	return nil
+}
+
+// audit audits resources periodically
+func (am *auditManager) audit() error {
+	c := am.mgr.GetClient()
+
+	// don't audit anything until the constraintTemplate crd is in the cluster
+	crd := &apiextensionsv1beta1.CustomResourceDefinition{}
+	if err := c.Get(context.TODO(), types.NamespacedName{Name: crdName}, crd); err != nil {
```

Comment:
Nit: should we pass the ctx from the auditManagerLoop here instead of using context.TODO?